### PR TITLE
Smarter PacketDispatcher

### PR DESCRIPTION
**Pull request by pveentjer**
All the dispatching is now done in the PacketDispatcher and removed from the OperationService. A layer less to jump through.

### CODE REVIEW

**Comment by pveentjer** [comment]
run-lab-run

**Comment by jerrinot** [comment]
:+1: good one!

**Comment by pveentjer** [comment]
run-lab-run

**Comment by pveentjer** [comment]
verify

File: hazelcast/src/main/java/com/hazelcast/spi/impl/packetdispatcher/impl/PacketDispatcherImpl.java [hunk 1]
```diff
@@ -25,42 +25,51 @@
...
     }
 
     @Override
     public void dispatch(Packet packet) {
         try {
-            if (packet.isFlagSet(FLAG_OP)) {
-                operationPacketHandler.handle(packet);
-            } else if (packet.isFlagSet(FLAG_EVENT)) {
+            int flags = packet.getFlags();
+            if ((flags & FLAG_OP) != 0) {
```
Comment (Donnerbart) [inline hazelcast/src/main/java/com/hazelcast/spi/impl/packetdispatcher/impl/PacketDispatcherImpl.java]: Hmm, this looks like a micro-optimization to me. I think `packet.isFlagSet()` is easier to read and should anyway get inlined by the JVM. It's also nearly the same amount of characters and less operators to read.

Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/spi/impl/packetdispatcher/impl/PacketDispatcherImpl.java]: I cleaned up the flags to be able do smarter bitlogic outside of these operations. Personally I find this easier to read.

Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/spi/impl/packetdispatcher/impl/PacketDispatcherImpl.java]: In c you could even simplify it to:

```
 @Override
    public void dispatch(Packet packet) {
        try {
            int flags = packet.getFlags();
            if (flags & FLAG_OP) {
                if (flags & FLAG_RESPONSE) {
                    responseHandler.handle(packet);
                } else {
                    operationExecutor.handle(packet);
                }
            } else if (flags & FLAG_EVENT) {
                eventPacketHandler.handle(packet);
            } else if (flags & FLAG_BIND) {
                connectionPacketHandler.handle(packet);
            } else {
                logger.severe("Unknown packet type! Header: " + packet.getFlags());
            }
        } catch (Throwable t) {
            inspectOutputMemoryError(t);
            logger.severe("Failed to process:" + packet, t);
        }
    }
```

Comment (mdogan) [inline hazelcast/src/main/java/com/hazelcast/spi/impl/packetdispatcher/impl/PacketDispatcherImpl.java]: `packet.isFlagSet()` is just plain English. how come `(flags & FLAG_OP) != 0` is easier to read?

Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/spi/impl/packetdispatcher/impl/PacketDispatcherImpl.java]: Perhaps customs? In my c days, that is how you normally work with flags.

Comment (mdogan) [inline hazelcast/src/main/java/com/hazelcast/spi/impl/packetdispatcher/impl/PacketDispatcherImpl.java]: In someone's assembly days, it was done like;

```
movl $0x1, %edx
movl $0x0, %ecx
andl %edx, %ecx
...
```

Comment (Donnerbart) [inline hazelcast/src/main/java/com/hazelcast/spi/impl/packetdispatcher/impl/PacketDispatcherImpl.java]: I like the idea of combining flags outside of `Packet`, but keeping the bit logic inside the class. So the counterpart of the very nice `packet.setAllFlags(FLAG_A | FLAG_B)` would be something like this:
```
if (packet.hasAllFlags(FLAG_OP | FLAG_RESPONSE)) {
    responseHandler.handle(packet);
} else if (packet.hasAllFlags(FLAG_OP)) {
    operationExecutor.handle(packet);
} else if (packet.hasAllFlags(FLAG_EVENT)) {
```

Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/spi/impl/packetdispatcher/impl/PacketDispatcherImpl.java]: Personally I really don't see the added value of repeatedly asking the packet for the same data and to check if some bits are set. Perhaps it is a premature optimization, perhaps it is just clean code. So this is a flavor based issue and we'll not reach consensus on this.

Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/spi/impl/packetdispatcher/impl/PacketDispatcherImpl.java]: I case of doubt I look at Aeron/Agrona:

This comes from Aeron controlled fragment assembler:

```
    public Action onFragment(final DirectBuffer buffer, final int offset, final int length, final Header header)
    {
        final byte flags = header.flags();

        Action action = Action.CONTINUE;

        if ((flags & UNFRAGMENTED) == UNFRAGMENTED)
        {
            action = delegate.onFragment(buffer, offset, length, header);
        }
        else
        {
            if ((flags & BEGIN_FRAG_FLAG) == BEGIN_FRAG_FLAG)
            {
                final BufferBuilder builder = builderBySessionIdMap.computeIfAbsent(header.sessionId(), builderFunc);
                builder.reset().append(buffer, offset, length);
            }
            else
```

Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/spi/impl/packetdispatcher/impl/PacketDispatcherImpl.java]: What about this one:

```
  @Override
    public void dispatch(Packet packet) {
        try {
            int flags = packet.getFlags() & (FLAG_OP | FLAG_RESPONSE | FLAG_EVENT | FLAG_BIND);
            switch (flags) {
                case FLAG_OP:
                    operationExecutor.handle(packet);
                    break;
                case FLAG_OP | FLAG_RESPONSE:
                    responseHandler.handle(packet);
                    break;
                case FLAG_EVENT:
                    eventPacketHandler.handle(packet);
                    break;
                case FLAG_BIND:
                    connectionPacketHandler.handle(packet);
                    break;
                default:
                    logger.severe("Unknown packet type! Header: " + packet.getFlags());
            }
        } catch (Throwable t) {
            inspectOutputMemoryError(t);
            logger.severe("Failed to process:" + packet, t);
        }
    }
```

Comment (noctarius) [inline hazelcast/src/main/java/com/hazelcast/spi/impl/packetdispatcher/impl/PacketDispatcherImpl.java]: I personally would agree with Mehmet and David :) Anyhow my solution would be different:

```
enum PacketFlags {
  FLAG_X;

  static boolean flagSet(int flags, PacketFlags flag) {
    return flags & flag == flag;
  }
}
```

**Comment by pveentjer** [comment]
Unless someone has a non opinion based reason to get this pr merged, I'm going to merge this one.

**Comment by noctarius** [comment]
:+1: no other objections :D

**Comment by Donnerbart** [comment]
:+1: Good cleanup :)

**Pull request merged**